compiler/rustc_driver/src/lib.rs

```diff
@@ -1151,23 +1151,26 @@ pub fn catch_with_exit_code(f: impl FnOnce() -> interface::Result<()>) -> i32 {
 static DEFAULT_HOOK: SyncLazy<Box<dyn Fn(&panic::PanicInfo<'_>) + Sync + Send + 'static>> =
     SyncLazy::new(|| {
         let hook = panic::take_hook();
-        panic::set_hook(Box::new(|info| report_ice(info, BUG_REPORT_URL)));
+        panic::set_hook(Box::new(|info| {
+            // Invoke the default handler, which prints the actual panic message and optionally a backtrace
+            (*DEFAULT_HOOK)(info);
+
+            // Separate the output with an empty line
+            eprintln!();
+
+            // Print the ICE message
+            report_ice(info, BUG_REPORT_URL);
+        }));
```

Comment:
This makes `report_ice` useless for the situations without enough control over the driver to bypass the `SyncLazy::force(&DEFAULT_HOOK);`. Maybe there should be a `report_ice_including_default_panic_hook` function that provides the old behavior?

Though ideally the new behavior should probably use a new function name (or better, both should change names, so use sites have to pick).

Reply:
You can set a panic hook more than once, so you can first get the default hook, then let whatever code forces `DEFAULT_HOOK` run and then set the new panic hook.

Reply:
A `report_ice_including_default_panic_hook` function requires the `DEFAULT_HOOK` to be forced first to be able to call it without ending in an infinite loop.

Reply:
> so you can first get the default hook

The codegen backend is loaded *after* `DEFAULT_HOOK` is initially forced, so it's impossible to make it work without changes to `rustc_driver`. I'll hopefully take a look at it soon and try to come up with a better overall solution. It should be possible to e.g. capture more data into the boxed closure, and the use of a `static` is somewhat incidental.

Though the most direct solution is to offer a `static` that *contains* the report URL (and maybe additional e.g. footer), and which can be changed independently of the panic hook.

Reply:
You could have something like https://github.com/bjorn3/rustc_codegen_cranelift/blob/d498e6d6971575714353f11aa4d3e63a9d2030b2/src/bin/cg_clif.rs#L20-L34 except that you call `panic::take_hook()` twice and use the result of the second time as default hook. The first time you take the hook set by rustc_driver and ignore it. The second time you take the default libstd hook.

Reply:
> If no custom hook is registered, the default hook will be returned.

Ah, this behavior documented for `std::panic::take_hook` was what I missed, thanks!